### PR TITLE
Todo list challenge

### PR DESCRIPTION
**Pull request by natstar93**

### CODE REVIEW

**Comment by natstar93** [comment]
Three questions to focus on in AH review:
- Form validation. Form is not reset to pristine on task addition
- MongoDB questions
- My HTML page is huge, but my controller is tiny. Is this ok?

(Also, my build is currently failing due to bower install failing, but I think a solution to this has been posted. Will fix before coach review)

File: index.html [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    <script src="js/todoController.js"></script>
+  </head>
+
+  <body ng-controller="TodoController as todoCtrl">
+    <div class="col-md-6 col-md-offset-3">
+      <h1>To Do List of Joy</h1>
+      <form class="form-horizontal task-form" name='addForm' ng-submit='todoCtrl.addTask()' novalidate>
+        <input type="text" class="col-md-6" name="taskDescription" ng-model="todoCtrl.taskDescription" ng-minlength="2" required>
+        <button class="btn btn-success add-task-btn" ng-click="addForm.taskDescription.$setPristine()">Add task</button>
+        <span style="color:red" ng-show="addForm.taskDescription.$dirty && addForm.taskDescription.$invalid" class="error-msg">
```
Comment (ptolemybarnes) [inline index.html:21]: Try not to have styling inline. Have it in a stylesheet.

File: js/todoController.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+todo.controller('TodoController', [function() {
+
```
Comment (ptolemybarnes) [inline js/todoController.js:2]: Random line breaks >_<

File: js/todoController.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  }
+
+  self.addTask = function() {
+    if (self.taskDescription) {
+      self.taskList.items.push( { "description" : self.taskDescription, "completed": false } );
+    }
+    self.taskDescription = '';
+  };
+
+  self.deleteCompleted = function() {
```
Comment (ptolemybarnes) [inline js/todoController.js:17]: Hmm what does 'deleteCompleted' mean?

File: test/e2e/todoFeature.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
+describe('To do list', function() {
+
+  var descriptionBox = element(by.model('todoCtrl.taskDescription'));
```
Comment (ptolemybarnes) [inline test/e2e/todoFeature.js:3]: Argh this is a little gnarly. Not sure how to refactor it though. >_<

File: test/e2e/todoFeature.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  describe('adding tasks', function(){
+    it('can add a new task', function() {
+      descriptionBox.sendKeys('Write feature test');
+      addTaskButton.click();
+      expect(element.all(by.binding('task.description')).last().getText()).toContain('Write feature test');
+    });
+
+    it('does not add task if description too short', function() {
+      descriptionBox.sendKeys('W');
+      expect(errorBox.element(by.className('length-error')).getText()).toEqual('Description must be 2 or more characters long.')
```
Comment (ptolemybarnes) [inline test/e2e/todoFeature.js:36]: try not to have multiple expectations in unit tests. you might want a different test for the error message

File: test/e2e/todoFeature.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+      addTaskButton.click();
+      expect(browser.isElementPresent(by.binding('task.description'))).toBe(false);
+    });    
+
+    it('does not add task if description missing', function() {
+      descriptionBox.sendKeys(' ');
```
Comment (ptolemybarnes) [inline test/e2e/todoFeature.js:42]: You've got sendKeys followed by click quite a few times. Could this be refactored?

File: test/e2e/todoFeature.js [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+  beforeEach(function() {
+    browser.get('http://localhost:8080');
+  });
+
+  it('has a title', function() {
+    expect(browser.getTitle()).toEqual('Todo List of Joy');
+  });
+
+  createTask = function() {
```
Comment (ptolemybarnes) [inline test/e2e/todoFeature.js:22]: Global variable >_<

File: test/e2e/todoFeature.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+    it('does not add task if description missing', function() {
+      descriptionBox.sendKeys(' ');
+      expect(errorBox.element(by.className('required-error')).getText()).toEqual('Description is required.')
+      addTaskButton.click();
+      expect(browser.isElementPresent(by.binding('task.description'))).toBe(false);
+    });  
+  });
+
+  it('can edit task', function() {
+    createTask();
```
Comment (ptolemybarnes) [inline test/e2e/todoFeature.js:50]: This is a nice refactoring but it's not very transparent. What task am I creating? It means that on line 62 the text 'Buy Vietnamese food' is a bit random - why am i expecting that text? Might be better if the creatTask function took the name of the task you want to create as a parameter.

File: test/e2e/todoFeature.js [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+    completedTick.click();
+    deleteCompletedBtn.click();
+    expect(element.all(by.binding('task.description')).getText()).toNotContain('Buy Vietnamese food');
+  });
+
+  it('shows total number of tasks', function() {
+    expect(element(by.className('total-tasks')).getText()).toContain('Total tasks: 0');
+    createTask();
+    expect(element(by.className('total-tasks')).getText()).toContain('Total tasks: 1');
```
Comment (ptolemybarnes) [inline test/e2e/todoFeature.js:81]: Again, try not to have multiple expectations in unit tests.

**Comment by ptolemybarnes** [comment]
Pretty good! Have a look at my suggested refactorings but otherwise, nice one :+1: :checkered_flag:

**Pull request closed without merging**